utils/src/main/java/software/amazon/awssdk/utils/http/SdkHttpUtils.java

```diff
@@ -322,4 +328,15 @@ public static String appendUri(String baseUri, String path) {
     public static boolean isSingleHeader(String h) {
         return SINGLE_HEADERS.contains(StringUtils.lowerCase(h));
     }
+
+    /**
+     * Extracts query parameters from the given URI
+     */
+    public static Map<String, List<String>> uriParams(URI uri) {
+        return Pattern.compile(QUERY_PARAM_PATTERN)
+               .splitAsStream(uri.getRawQuery().trim())
+               .map(s -> s.contains("=") ? s.split("=", 2) : new String[] {s, ""})
```

Comment:
Hmm I don't think using an empty string when there's no `=` is what we want; we should be setting `null` if there's no `=` for the parameter. There should be a test for the result of something like this is

```
URI myUri = URI.create("https://github.com/aws/aws-sdk-for-java-v2?foo");

SdkHttpFullRequest request = SdkHttpFullRequest.builder().uri(myUri).build();

request.getUri();
```

to make sure that the result is not `https://github.com/aws/aws-sdk-for-java-v2?foo=`, which will likely be interpreted differently by the server